mongo/integration/unified/entity_test.go

```diff
@@ -0,0 +1,294 @@
+// ID, even if they are of different types. It also enforces referential integrity so construction of an entity that
+// references another (e.g. a database entity references a client) will fail if the referenced entity does not exist.
+type EntityMap struct {
+	allEntities   map[string]struct{}
+	changeStreams map[string]*mongo.ChangeStream
+	clients       map[string]*ClientEntity
+	dbs           map[string]*mongo.Database
+	collections   map[string]*mongo.Collection
+	sessions      map[string]mongo.Session
+	gridfsBuckets map[string]*gridfs.Bucket
+	bsonValues    map[string]bson.RawValue
+}
+
+func NewEntityMap() *EntityMap {
+	return &EntityMap{
+		allEntities:   make(map[string]struct{}),
+		gridfsBuckets: make(map[string]*gridfs.Bucket),
+		bsonValues:    make(map[string]bson.RawValue),
+		changeStreams: make(map[string]*mongo.ChangeStream),
+		clients:       make(map[string]*ClientEntity),
+		collections:   make(map[string]*mongo.Collection),
+		dbs:           make(map[string]*mongo.Database),
+		sessions:      make(map[string]mongo.Session),
+	}
+}
+
+func (em *EntityMap) AddBSONEntity(id string, val bson.RawValue) error {
+	if err := em.verifyEntityDoesNotExist(id); err != nil {
+		return err
+	}
+
+	em.allEntities[id] = struct{}{}
+	em.bsonValues[id] = val
+	return nil
+}
+
+func (em *EntityMap) AddChangeStreamEntity(id string, stream *mongo.ChangeStream) error {
+	if err := em.verifyEntityDoesNotExist(id); err != nil {
+		return err
+	}
+
+	em.allEntities[id] = struct{}{}
+	em.changeStreams[id] = stream
+	return nil
+}
+
+func (em *EntityMap) AddEntity(ctx context.Context, entityType string, entityOptions *EntityOptions) error {
+	if err := em.verifyEntityDoesNotExist(entityOptions.ID); err != nil {
+		return err
+	}
+
+	var err error
+	switch entityType {
+	case "client":
+		err = em.addClientEntity(ctx, entityOptions)
+	case "database":
+		err = em.addDatabaseEntity(entityOptions)
+	case "collection":
+		err = em.addCollectionEntity(entityOptions)
+	case "session":
+		err = em.addSessionEntity(entityOptions)
+	case "bucket":
+		err = em.addGridFSBucketEntity(entityOptions)
+	default:
+		return fmt.Errorf("unrecognized entity type %q", entityType)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error constructing entity of type %q: %v", entityType, err)
+	}
+	em.allEntities[entityOptions.ID] = struct{}{}
+	return nil
+}
+
+func (em *EntityMap) GridFSBucket(id string) (*gridfs.Bucket, error) {
+	bucket, ok := em.gridfsBuckets[id]
+	if !ok {
+		return nil, newEntityNotFoundError("gridfs bucket", id)
+	}
+	return bucket, nil
+}
+
+func (em *EntityMap) BSONValue(id string) (bson.RawValue, error) {
+	val, ok := em.bsonValues[id]
+	if !ok {
+		return emptyRawValue, newEntityNotFoundError("BSON", id)
+	}
+	return val, nil
+}
+
+func (em *EntityMap) ChangeStream(id string) (*mongo.ChangeStream, error) {
+	client, ok := em.changeStreams[id]
+	if !ok {
+		return nil, newEntityNotFoundError("client", id)
```

Comment:
```suggestion
		return nil, newEntityNotFoundError("changestream", id)
```

Reply:
Done with a small modification: "change stream" instead of "changestream".